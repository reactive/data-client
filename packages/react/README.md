<h1>
<div align="center">
<a href="https://dataclient.io" target="_blank" rel="noopener">
  <img alt="Reactive Data Client" src="./data_client_logo_and_text.svg?sanitize=true">
</a>
</div>
</h1>

The scalable way to build applications with [dynamic data](https://dataclient.io/docs/getting-started/mutations).

[Declarative resouce definitons](https://dataclient.io/docs/getting-started/resource) for [REST](https://dataclient.io/rest), [GraphQL](https://dataclient.io/graphql), [Websockets+SSE](https://dataclient.io/docs/concepts/managers#data-stream) and [more](https://dataclient.io/rest/api/Endpoint)
<br/>[Performant rendering](https://dataclient.io/docs/getting-started/data-dependency) in [React](https://react.dev/), [NextJS](https://nextjs.org/), [React Native](https://reactnative.dev/), [Expo](https://expo.dev/)

Schema driven. Zero updater functions.

<div align="center">

[![CircleCI](https://circleci.com/gh/reactive/data-client/tree/master.svg?style=shield)](https://circleci.com/gh/reactive/data-client)
[![Coverage Status](https://img.shields.io/codecov/c/gh/reactive/data-client/master.svg?style=flat-square)](https://app.codecov.io/gh/reactive/data-client?branch=master)
[![Percentage of issues still open](https://isitmaintained.com/badge/open/reactive/data-client.svg)](https://github.com/reactive/data-client/issues 'Percentage of issues still open')
[![bundle size](https://img.shields.io/bundlephobia/minzip/@data-client/react?style=flat-square)](https://bundlephobia.com/result?p=@data-client/react)
[![npm version](https://img.shields.io/npm/v/@data-client/react.svg?style=flat-square)](https://www.npmjs.com/package/@data-client/react)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
[![Chat](https://img.shields.io/discord/768254430381735967.svg?style=flat-square&colorB=758ED3)](https://discord.gg/35nb8Mz)

**[📖Read The Docs](https://dataclient.io/docs)** &nbsp;|&nbsp; [🏁Getting Started](https://dataclient.io/docs/getting-started/installation) <br/>🎮 **Demos:** &nbsp;
[Todo](https://stackblitz.com/github/reactive/data-client/tree/master/examples/todo-app?file=src%2Fpages%2FHome%2FTodoList.tsx) &nbsp;|&nbsp;
[Github](https://stackblitz.com/github/reactive/data-client/tree/master/examples/github-app?file=src%2Fpages%2FIssueList.tsx) &nbsp;|&nbsp;
[NextJS SSR](https://stackblitz.com/github/reactive/data-client/tree/master/examples/nextjs?file=components%2Ftodo%2FTodoList.tsx) &nbsp;|&nbsp;
[Websockets+SSR](https://coin-app-lake.vercel.app)

</div>

## Installation

```bash
npm install --save @data-client/react @data-client/rest @data-client/test
```

For more details, see [the Installation docs page](https://dataclient.io/docs/getting-started/installation).

## Usage

### Simple [TypeScript definition](https://dataclient.io/rest/api/Entity)

```typescript
class User extends Entity {
  id = '';
  username = '';
}

class Article extends Entity {
  id = '';
  title = '';
  body = '';
  author = User.fromJS();
  createdAt = Temporal.Instant.fromEpochMilliseconds(0);

  static schema = {
    author: User,
    createdAt: Temporal.Instant.from,
  };
}
```

### Create [collection of API Endpoints](https://dataclient.io/docs/getting-started/resource)

```typescript
const UserResource = resource({
  path: '/users/:id',
  schema: User,
  optimistic: true,
});

const ArticleResource = resource({
  path: '/articles/:id',
  schema: Article,
  searchParams: {} as { author?: string },
  optimistic: true,
  paginationField: 'cursor',
});
```

### One line [data binding](https://dataclient.io/docs/getting-started/data-dependency)

```tsx
const article = useSuspense(ArticleResource.get, { id });
return (
  <article>
    <h2>
      {article.title} by {article.author.username}
    </h2>
    <p>{article.body}</p>
  </article>
);
```

### [Reactive Mutations](https://dataclient.io/docs/getting-started/mutations)

```tsx
const ctrl = useController();
return (
  <>
    <CreateArticleForm
      onSubmit={article =>
        ctrl.fetch(ArticleResource.getList.push, { id }, article)
      }
    />
    <ProfileForm
      onSubmit={user =>
        ctrl.fetch(UserResource.update, { id: article.author.id }, user)
      }
    />
    <button onClick={() => ctrl.fetch(ArticleResource.delete, { id })}>
      Delete
    </button>
  </>
);
```

### [Subscriptions](https://dataclient.io/docs/api/useLive)

```tsx
const price = useLive(PriceResource.get, { symbol });
return price.value;
```

### [Type-safe Imperative Actions](https://dataclient.io/docs/api/Controller)

```tsx
const ctrl = useController();
await ctrl.fetch(ArticleResource.update, { id }, articleData);
await ctrl.fetchIfStale(ArticleResource.get, { id });
ctrl.expireAll(ArticleResource.getList);
ctrl.invalidate(ArticleResource.get, { id });
ctrl.invalidateAll(ArticleResource.getList);
ctrl.setResponse(ArticleResource.get, { id }, articleData);
ctrl.set(Article, { id }, articleData);
```

### [Programmatic queries](https://dataclient.io/rest/api/Query)

```typescript
const queryTotalVotes = new schema.Query(
  new schema.Collection([BlogPost]),
  posts => posts.reduce((total, post) => total + post.votes, 0),
);

const totalVotes = useQuery(queryTotalVotes);
const totalVotesForUser = useQuery(queryTotalVotes, { userId });
```

```typescript
const groupTodoByUser = new schema.Query(
  TodoResource.getList.schema,
  todos => Object.groupBy(todos, todo => todo.userId),
);
const todosByUser = useQuery(groupTodoByUser);
```

### [Powerful Middlewares](https://dataclient.io/docs/concepts/managers)

```ts
class LoggingManager implements Manager {
  middleware: Middleware = controller => next => async action => {
    console.log('before', action, controller.getState());
    await next(action);
    console.log('after', action, controller.getState());
  };

  cleanup() {}
}
```

```ts
class TickerStream implements Manager {
  middleware: Middleware = controller => {
    this.handleMsg = msg => {
      controller.set(Ticker, { id: msg.id }, msg);
    };
    return next => action => next(action);
  };

  init() {
    this.websocket = new WebSocket('wss://ws-feed.myexchange.com');
    this.websocket.onmessage = event => {
      const msg = JSON.parse(event.data);
      this.handleMsg(msg);
    };
  }
  cleanup() {
    this.websocket.close();
  }
}
```

### [Integrated data mocking](https://dataclient.io/docs/api/Fixtures)

```tsx
const fixtures = [
  {
    endpoint: ArticleResource.getList,
    args: [{ maxResults: 10 }] as const,
    response: [
      {
        id: '5',
        title: 'first post',
        body: 'have a merry christmas',
        author: { id: '10', username: 'bob' },
        createdAt: new Date(0).toISOString(),
      },
      {
        id: '532',
        title: 'second post',
        body: 'never again',
        author: { id: '10', username: 'bob' },
        createdAt: new Date(0).toISOString(),
      },
    ],
  },
  {
    endpoint: ArticleResource.update,
    response: ({ id }, body) => ({
      ...body,
      id,
    }),
  },
];

const Story = () => (
  <MockResolver fixtures={options[result]}>
    <ArticleList maxResults={10} />
  </MockResolver>
);
```

### ...all typed ...fast ...and consistent

For the small price of 9kb gziped. &nbsp;&nbsp; [🏁Get started now](https://dataclient.io/docs/getting-started/installation)

## Features

- [x] ![TS](./typescript.svg?sanitize=true) Strong [Typescript](https://www.typescriptlang.org/) inference
- [x] 🛌 React [Suspense](https://dataclient.io/docs/getting-started/data-dependency#boundaries) support
- [x] 🧵 React 18 [Concurrent mode](https://dataclient.io/docs/guides/render-as-you-fetch) compatible
- [x] 💦 [Partial Hydration Server Side Rendering](https://dataclient.io/docs/guides/ssr)
- [x] 🎣 [Declarative API](https://dataclient.io/docs/getting-started/data-dependency)
- [x] 📝 Composition over configuration
- [x] 💰 [Normalized](https://dataclient.io/docs/concepts/normalization) caching
- [x] 💥 Tiny bundle footprint
- [x] 🛑 Automatic overfetching elimination
- [x] ✨ Fast [optimistic updates](https://dataclient.io/rest/guides/optimistic-updates)
- [x] 🧘 [Flexible](https://dataclient.io/docs/getting-started/resource) to fit any API design (one size fits all)
- [x] 🔧 [Debugging and inspection](https://dataclient.io/docs/getting-started/debugging) via browser extension
- [x] 🌳 Tree-shakable (only use what you need)
- [x] 🔁 [Subscriptions](https://dataclient.io/docs/api/useSubscription)
- [x] ♻️ Optional [redux integration](https://dataclient.io/docs/guides/redux)
- [x] 📙 [Storybook mocking](https://dataclient.io/docs/guides/storybook)
- [x] 📱 [React Native](https://facebook.github.io/react-native/) support
- [x] 📱 [Expo](https://dataclient.io/docs/getting-started/installation) support
- [x] ⚛️ [NextJS](https://dataclient.io/docs/guides/ssr#nextjs) support
- [x] 🚯 [Declarative cache lifetime policy](https://dataclient.io/docs/concepts/expiry-policy)
- [x] 🧅 [Composable middlewares](https://dataclient.io/docs/api/Manager)
- [x] 💽 Global data consistency guarantees
- [x] 🏇 Automatic race condition elimination
- [x] 👯 Global referential equality guarantees

## Examples

- Todo: [![GitHub](https://badgen.net/badge/icon/github?icon=github&label)](https://github.com/reactive/data-client/tree/master/examples/todo-app) | [![Sandbox](https://developer.stackblitz.com/img/open_in_stackblitz_small.svg)](https://stackblitz.com/github/reactive/data-client/tree/master/examples/todo-app?file=src%2Fpages%2FHome%2FTodoList.tsx)
- Github: [![GitHub](https://badgen.net/badge/icon/github?icon=github&label)](https://github.com/reactive/data-client/tree/master/examples/github-app) | [![Sandbox](https://developer.stackblitz.com/img/open_in_stackblitz_small.svg)](https://stackblitz.com/github/reactive/data-client/tree/master/examples/github-app?file=src%2Fpages%2FIssueList.tsx)
- NextJS: [![GitHub](https://badgen.net/badge/icon/github?icon=github&label)](https://github.com/reactive/data-client/tree/master/examples/nextjs) | [![Sandbox](https://developer.stackblitz.com/img/open_in_stackblitz_small.svg)](https://stackblitz.com/github/reactive/data-client/tree/master/examples/nextjs?file=components%2Ftodo%2FTodoList.tsx)
- Websockets: [![GitHub](https://badgen.net/badge/icon/github?icon=github&label)](https://github.com/reactive/coin-app) | [![Sandbox](https://developer.stackblitz.com/img/open_in_stackblitz_small.svg)](https://stackblitz.com/github/reactive/coin-app/tree/master?file=src%2Fresources%2FStreamManager.ts,src%2Fapp%2FCurrencyList.tsx) | [![Website](https://badgen.net/badge/icon/production?icon=vercel&label)](https://coin-app-lake.vercel.app)

## API

- Rendering: [useSuspense()](https://dataclient.io/docs/api/useSuspense), [useLive()](https://dataclient.io/docs/api/useLive), [useCache()](https://dataclient.io/docs/api/useCache), [useDLE()](https://dataclient.io/docs/api/useDLE), [useQuery()](https://dataclient.io/docs/api/useQuery), [useLoading()](https://dataclient.io/docs/api/useLoading), [useDebounce()](https://dataclient.io/docs/api/useDebounce), [useCancelling()](https://dataclient.io/docs/api/useCancelling)
- Event handling: [useController()](https://dataclient.io/docs/api/useController) returns [Controller](https://dataclient.io/docs/api/Controller)
  - [ctrl.fetch](https://dataclient.io/docs/api/Controller#fetch)
  - [ctrl.fetchIfStale](https://dataclient.io/docs/api/Controller#fetchIfStale)
  - [ctrl.expireAll](https://dataclient.io/docs/api/Controller#expireAll)
  - [ctrl.invalidate](https://dataclient.io/docs/api/Controller#invalidate)
  - [ctrl.invalidateAll](https://dataclient.io/docs/api/Controller#invalidateAll)
  - [ctrl.resetEntireStore](https://dataclient.io/docs/api/Controller#resetEntireStore)
  - [ctrl.set](https://dataclient.io/docs/api/Controller#set)
  - [ctrl.setResponse](https://dataclient.io/docs/api/Controller#setResponse)
  - [ctrl.setError](https://dataclient.io/docs/api/Controller#setError)
  - [ctrl.resolve](https://dataclient.io/docs/api/Controller#resolve)
  - [ctrl.subscribe](https://dataclient.io/docs/api/Controller#subscribe)
  - [ctrl.unsubscribe](https://dataclient.io/docs/api/Controller#unsubscribe)
- Components: [&lt;DataProvider/>](https://dataclient.io/docs/api/DataProvider), [&lt;AsyncBoundary/>](https://dataclient.io/docs/api/AsyncBoundary), [&lt;ErrorBoundary/>](https://dataclient.io/docs/api/ErrorBoundary)
- Middleware: [LogoutManager](https://dataclient.io/docs/api/LogoutManager), [NetworkManager](https://dataclient.io/docs/api/NetworkManager), [SubscriptionManager](https://dataclient.io/docs/api/SubscriptionManager), [PollingSubscription](https://dataclient.io/docs/api/PollingSubscription), [DevToolsManager](https://dataclient.io/docs/api/DevToolsManager)
