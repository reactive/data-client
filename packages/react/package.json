{
  "name": "@data-client/react",
  "version": "0.14.25",
  "description": "Async State Management without the Management. REST, GraphQL, SSE, Websockets, Fetch",
  "homepage": "https://dataclient.io",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/react"
  },
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "keywords": [
    "react",
    "data",
    "cache",
    "flux",
    "suspense",
    "fetch",
    "hook",
    "networking",
    "async",
    "concurrent mode",
    "typescript",
    "async",
    "data fetching",
    "data cache",
    "reactive",
    "state management",
    "api client",
    "api",
    "normalized cache",
    "swr",
    "query",
    "front-end",
    "mobile",
    "react-native",
    "ios",
    "android",
    "web",
    "expo",
    "expogo",
    "nextjs",
    "middleware",
    "websocket",
    "REST",
    "GraphQL",
    "RPC",
    "sse",
    "declarative",
    "dynamic data",
    "mutations"
  ],
  "sideEffects": [
    "./lib/hooks/useFocusEffect.native.js",
    "./native/hooks/useFocusEffect.native.js"
  ],
  "main": "dist/index.js",
  "react-native": "native/index.js",
  "module": "legacy/index.js",
  "unpkg": "dist/index.umd.min.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "next": [
        "lib/next/index.d.ts"
      ],
      "nextjs": [
        "lib/server/nextjs/index.d.ts"
      ],
      "ssr": [
        "lib/server/index.d.ts"
      ],
      "redux": [
        "lib/server/redux/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "next": [
        "ts3.4/next/index.d.ts"
      ],
      "ssr": [
        "ts3.4/server/index.d.ts"
      ],
      "redux": [
        "ts3.4/server/redux/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/index.js",
      "browser": "./lib/index.js",
      "react-native": "./native/index.js",
      "default": "./lib/index.js"
    },
    "./next": {
      "types": "./lib/next/index.d.ts",
      "require": "./dist/next.js",
      "browser": "./lib/next/index.js",
      "react-native": "./native/next/index.js",
      "default": "./lib/next/index.js"
    },
    "./nextjs": {
      "types": "./lib/server/nextjs/index.d.ts",
      "default": "./lib/server/nextjs/index.js"
    },
    "./ssr": {
      "types": "./lib/server/index.d.ts",
      "require": "./dist/server/index.js",
      "default": "./lib/server/index.js"
    },
    "./redux": {
      "types": "./lib/server/redux/index.d.ts",
      "require": "./dist/server/redux/index.js",
      "browser": "./lib/server/redux/index.js",
      "react-native": "./native/server/redux/index.js",
      "default": "./lib/server/redux/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "legacy",
    "native",
    "ts3.4",
    "node.mjs",
    "LICENSE",
    "README.md",
    "./data_client_logo_and_text.svg",
    "./typescript.svg"
  ],
  "scripts": {
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' POLYFILL_TARGETS='chrome>88,safari>14' POLYFILL_TARGETS='chrome>88,safari>14' yarn g:babel --out-dir lib",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' POLYFILL_TARGETS='chrome>80,safari>13,firefox>72' POLYFILL_TARGETS='chrome>80,safari>13,firefox>72' yarn g:babel --out-dir legacy",
    "build:native:lib": "COMPILE_TARGET=native NODE_ENV=production BROWSERSLIST_ENV='hermes12' yarn g:babel --out-dir native",
    "build:js:node": "BROWSERSLIST_ENV=node12 yarn g:rollup",
    "build:js:browser": "BROWSERSLIST_ENV=legacy yarn g:rollup",
    "build:bundle": "yarn g:runs build:js:\\* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "yarn g:clean index.d.ts next.d.ts native",
    "build:legacy-types": "yarn g:downtypes lib ts3.4",
    "build": "run build:lib && run build:legacy:lib && run build:native:lib && run build:bundle",
    "dev": "run build:lib -w",
    "prepare": "run build:lib",
    "prepack": "run prepare && run build:legacy:lib && run build:bundle"
  },
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "funding": "https://github.com/sponsors/ntucker",
  "license": "Apache-2.0",
  "dependencies": {
    "@babel/runtime": "^7.20.0",
    "@data-client/core": "^0.14.24",
    "@data-client/use-enhanced-reducer": "^0.1.14"
  },
  "peerDependencies": {
    "@react-navigation/native": "^6.0.0 || ^7.0.0",
    "@types/react": "^16.14.0 || ^17.0.0 || ^18.0.0 || ^19.0.0-0 || ^19.0.0",
    "react": "^16.14.0 || ^17.0.0 || ^18.0.0 || ^19.0.0-0 || ^19.0.0"
  },
  "peerDependenciesMeta": {
    "@react-navigation/native": {
      "optional": true
    },
    "@types/react": {
      "optional": true
    }
  },
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@data-client/rest": "workspace:*",
    "@data-client/test": "workspace:*",
    "@jest/globals": "^29.7.0",
    "@js-temporal/polyfill": "^0.5.0",
    "@react-navigation/native": "^7.0.0",
    "@react-navigation/native-stack": "^7.2.0",
    "@testing-library/dom": "^10.4.0",
    "@testing-library/react": "^16.1.0",
    "@testing-library/react-native": "^13.0.0",
    "@types/jest": "^29.5.14",
    "@types/node": "^22.0.0",
    "@types/qs": "^6",
    "@types/react": "19.1.6",
    "@types/react-dom": "*",
    "jest-mock": "^29.7.0",
    "nock": "13.3.1",
    "qs": "^6.13.1",
    "react": "19.1.0",
    "react-dom": "^19.0.0",
    "react-native": "0.79.3",
    "react-native-safe-area-context": "^5.0.0",
    "react-native-screens": "^4.1.0",
    "react-test-renderer": "^19.0.0",
    "rollup-plugins": "workspace:*"
  }
}
