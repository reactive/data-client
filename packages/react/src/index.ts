Object.hasOwn =
  Object.hasOwn ||
  /* istanbul ignore next */ function hasOwn(it, key) {
    return Object.prototype.hasOwnProperty.call(it, key);
  };
export { Controller, ExpiryStatus, actionTypes } from '@data-client/core';
export type {
  EndpointExtraOptions,
  FetchFunction,
  ResolveType,
  EndpointInterface,
  EntityInterface,
  Queryable,
  SchemaArgs,
  Schema,
  SchemaClass,
  DenormalizeNullable,
  Denormalize,
  Normalize,
  NormalizeNullable,
  FetchAction,
  InvalidateAction,
  UnsubscribeAction,
  SubscribeAction,
  ResetAction,
  SetAction,
  SetResponseAction,
  ActionTypes,
  NetworkError,
  UnknownError,
  ErrorTypes,
  AbstractInstanceType,
  UpdateFunction,
  State,
  PK,
  Dispatch,
  Middleware,
  MiddlewareAPI,
  Manager,
  GCInterface,
  GCOptions,
  CreateCountRef,
  // used in Controller generic
  DataClientDispatch,
  GenericDispatch,
} from '@data-client/core';
export { default as GCPolicy } from './state/GCPolicy.js';
export * from './managers/index.js';
export * from './components/index.js';
export * from './hooks/index.js';
export { StateContext, ControllerContext, StoreContext } from './context.js';
export type { Store } from './context.js';
export * as __INTERNAL__ from './internal.js';
export { usePromisifiedDispatch } from '@data-client/use-enhanced-reducer';
