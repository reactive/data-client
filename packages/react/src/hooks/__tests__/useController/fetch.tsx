import { ResolveType } from '@data-client/core';
import { CacheProvider } from '@data-client/react';
import { DataProvider as ExternalDataProvider } from '@data-client/react/redux';
import {
  CoolerArticle,
  CoolerArticleResource,
  FirstUnion,
  FutureArticleResource,
  UnionResource,
} from '__tests__/new';
import nock from 'nock';

import { useCache, useSuspense } from '../..';
// relative imports to avoid circular dependency in tsconfig references
import {
  makeRenderDataClient,
  FixtureEndpoint,
  act,
} from '../../../../../test';
import useController from '../../useController';

export const payload = {
  id: 5,
  title: 'hi ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};

export const createPayload = {
  id: 1,
  title: 'hi ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};

export const nested = [
  {
    id: 5,
    title: 'hi ho',
    content: 'whatever',
    tags: ['a', 'best', 'react'],
    author: {
      id: 23,
      username: 'bob',
    },
  },
  {
    id: 3,
    title: 'the next time',
    content: 'whatever',
    author: {
      id: 23,
      username: 'charles',
      email: 'bob@bob.com',
    },
  },
];

describe.each([
  ['CacheProvider', CacheProvider],
  ['ExternalDataProvider', ExternalDataProvider],
] as const)(`%s`, (_, makeProvider) => {
  // TODO: add nested resource test case that has multiple partials to test merge functionality

  let renderDataClient: ReturnType<typeof makeRenderDataClient>;
  let mynock: nock.Scope;

  beforeEach(() => {
    nock(/.*/)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      })
      .options(/.*/)
      .reply(200)
      .get(`/article-cooler/${payload.id}`)
      .reply(200, payload)
      .delete(`/article-cooler/${payload.id}`)
      .reply(204, '')
      .delete(`/article/${payload.id}`)
      .reply(200, {})
      .get(`/article-cooler/0`)
      .reply(403, {})
      .get(`/article-cooler/666`)
      .reply(200, '')
      .get(`/article-cooler`)
      .reply(200, nested)
      .post(`/article-cooler`)
      .reply(200, createPayload);

    mynock = nock(/.*/).defaultReplyHeaders({
      'Access-Control-Allow-Origin': '*',
      'Content-Type': 'application/json',
    });
  });

  afterEach(() => {
    nock.cleanAll();
  });

  beforeEach(() => {
    renderDataClient = makeRenderDataClient(makeProvider);
  });

  let errorspy: jest.Spied<any>;
  beforeEach(() => {
    errorspy = jest.spyOn(global.console, 'error').mockImplementation(() => {});
    jest.spyOn(global.console, 'warn').mockImplementation(() => {});
  });
  afterEach(() => {
    errorspy.mockRestore();
  });
  let warnSpy: jest.Spied<typeof console.warn>;
  afterEach(() => {
    warnSpy.mockRestore();
  });
  beforeEach(() => {
    (warnSpy = jest.spyOn(console, 'warn')).mockImplementation(() => {});
  });

  it('should fetch', async () => {
    const { result } = renderDataClient(() => {
      return {
        data: useCache(FutureArticleResource.get, payload.id),
        fetch: useController().fetch,
      };
    });
    expect(result.current.data).toBeUndefined();
    result.current.fetch(FutureArticleResource.get, payload.id);
    result.current.fetch(FutureArticleResource.get, payload.id);
    result.current.fetch(FutureArticleResource.get, payload.id);
    const response = await result.current.fetch(
      FutureArticleResource.get,
      payload.id,
    );
    expect(result.current.data).toBeDefined();
    expect(result.current.data?.content).toEqual(payload.content);
    expect(response).toEqual(CoolerArticle.fromJS(payload));
    expect(response.content).toBe(payload.content);

    // @ts-expect-error
    () => response.slkdf;

    // type tests
    // TODO: move these to own unit tests if/when applicable
    () => {
      // @ts-expect-error
      result.current.fetch(FutureArticleResource.get);
      // @ts-expect-error
      result.current.fetch(FutureArticleResource.get, {
        id: payload.id,
      });
      result.current.fetch(FutureArticleResource.create, payload);
      // @ts-expect-error
      result.current.fetch(FutureArticleResource.create, {}, payload);
      result.current.fetch(FutureArticleResource.update, payload.id, payload);
      // @ts-expect-error
      result.current.fetch(FutureArticleResource.update, payload);
    };
  });

  it('should fetch with resolver', async () => {
    // we use this id because it is not nock'd
    const id = 10000;
    const fixture = {
      endpoint: FutureArticleResource.get,
      args: [10000],
      response: payload,
    };
    const { result } = renderDataClient(
      () => {
        return {
          data: useCache(FutureArticleResource.get, id),
          fetch: useController().fetch,
        };
      },
      { resolverFixtures: [fixture] },
    );
    expect(result.current.data).toBeUndefined();
    result.current.fetch(FutureArticleResource.get, id);
    result.current.fetch(FutureArticleResource.get, id);
    result.current.fetch(FutureArticleResource.get, id);
    await result.current.fetch(FutureArticleResource.get, id);
  });

  it('should update on create', async () => {
    const endpoint = FutureArticleResource.create;
    const response: ResolveType<typeof endpoint> = createPayload;
    const fixtures: FixtureEndpoint[] = [
      {
        endpoint,
        args: [{ id: 1 }],
        response,
      },
    ];
    // use nock, and use resolver
    for (const resolverFixtures of [undefined, fixtures]) {
      const { result, waitForNextUpdate } = renderDataClient(
        () => {
          const articles = useSuspense(FutureArticleResource.getList);
          const fetch = useController().fetch;
          return { articles, fetch };
        },
        { resolverFixtures },
      );
      await waitForNextUpdate();
      await act(async () => {
        const article = await result.current.fetch(
          FutureArticleResource.create,
          {
            id: 1,
          },
        );
        article.content;
        // @ts-expect-error
        article.asdf;
        article.pk();
      });

      expect(result.current.articles.map(({ id }) => id)).toEqual([5, 3, 1]);
    }
  });

  it('should log error message when user update method throws', async () => {
    const endpoint = FutureArticleResource.create.extend({
      schema: CoolerArticle,
      update: () => {
        throw new Error('usererror');
      },
    });
    const response: ResolveType<typeof endpoint> = createPayload;
    const fixtures: FixtureEndpoint[] = [
      {
        endpoint,
        args: [{ id: 1 }],
        response,
      },
    ];
    const { result, waitForNextUpdate } = renderDataClient(
      () => {
        const articles = useSuspense(FutureArticleResource.getList);
        const fetch = useController().fetch;
        return { articles, fetch };
      },
      { resolverFixtures: fixtures },
    );
    await waitForNextUpdate();
    await act(async () => {
      await result.current.fetch(endpoint, {
        id: 1,
      });
    }),
      // still keeps old list
      expect(result.current.articles.map(({ id }) => id)).toEqual([5, 3]);

    expect(errorspy.mock.calls[0]).toMatchSnapshot();
  });

  it('should not suspend once deleted and redirected at same time', async () => {
    const temppayload = {
      ...payload,
      id: 1234,
    };
    mynock
      .get(`/article-cooler/${temppayload.id}`)
      .reply(200, temppayload)
      .delete(`/article-cooler/${temppayload.id}`)
      .reply(204, '');
    const throws: Promise<any>[] = [];
    const { result, waitForNextUpdate, rerender } = renderDataClient(
      ({ id }) => {
        try {
          return [
            useSuspense(FutureArticleResource.get, id) ?? null,
            useController().fetch,
          ] as const;
        } catch (e: any) {
          if (typeof e.then === 'function') {
            if (e !== throws[throws.length - 1]) {
              throws.push(e);
            }
          }
          throw e;
        }
      },
      { initialProps: { id: temppayload.id as number | null } },
    );
    expect(result.current).toBeUndefined();
    await waitForNextUpdate();
    const [data, fetch] = result.current;
    expect(data).toBeInstanceOf(CoolerArticle);
    expect(data?.title).toBe(temppayload.title);
    // react 19 suspends twice
    expect(throws.length).toBeGreaterThanOrEqual(1);

    mynock
      .persist()
      .get(`/article-cooler/${temppayload.id}`)
      .reply(200, { ...temppayload, title: 'othertitle' });

    // react 19 suspends twice
    expect(throws.length).toBeGreaterThanOrEqual(1);
    await act(async () => {
      await fetch(FutureArticleResource.delete, temppayload.id);
      rerender({ id: null });
    });
    // react 19 suspends twice
    expect(throws.length).toBeGreaterThanOrEqual(1);
    expect(result.current[0]).toBe(null);
  });

  it('should return denormalized value when schema is present', async () => {
    const { controller } = renderDataClient(
      () => {
        return 'hi';
      },
      {
        resolverFixtures: [
          {
            endpoint: CoolerArticleResource.get,
            args: [{ id: payload.id }],
            response: payload,
          },
        ],
      },
    );
    const ret = await controller.fetch(CoolerArticleResource.get, {
      id: payload.id,
    });
    expect(ret.content).toEqual(payload.content);
    expect(ret).toBeInstanceOf(CoolerArticle);
    expect(warnSpy.mock.calls.length).toBe(0);
  });

  it('should return denormalized value when schema is present (unions)', async () => {
    const response = [
      null,
      { id: '5', body: 'hi', type: 'first' },
      { id: '6', body: 'hi', type: 'another' },
      { id: '7', body: 'hi' },
    ];
    const { controller } = renderDataClient(
      () => {
        return 'hi';
      },
      {
        resolverFixtures: [
          {
            endpoint: UnionResource.getList,
            args: [],
            response,
          },
        ],
      },
    );
    const ret = await controller.fetch(UnionResource.getList);
    expect(ret[0]).toBeNull();
    expect(ret[1]).toBeInstanceOf(FirstUnion);
    expect(ret[2]).toEqual(response[2]);
    expect(ret[3]).toEqual(response[3]);
    expect(warnSpy.mock.calls).toMatchSnapshot();
  });
});
