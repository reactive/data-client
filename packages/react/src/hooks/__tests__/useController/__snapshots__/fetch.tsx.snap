// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CacheProvider should log error message when user update method throws 1`] = `
[
  "The following error occured during Endpoint.update() for POST http://test.com/article-cooler",
]
`;

exports[`CacheProvider should return denormalized value when schema is present (unions) 1`] = `
[
  [
    "Schema attribute "another" is not expected.
Expected one of: "first", "second"

Value: {
  "id": "6",
  "body": "hi",
  "type": "another"
}",
  ],
  [
    "Schema attribute undefined is not expected.
Expected one of: "first", "second"

Value: {
  "id": "7",
  "body": "hi"
}",
  ],
  [
    "TypeError: Unable to determine schema for UnionSchema
Value: {
  "id": "6",
  "body": "hi",
  "type": "another"
}.",
  ],
  [
    "TypeError: Unable to determine schema for UnionSchema
Value: {
  "id": "7",
  "body": "hi"
}.",
  ],
]
`;

exports[`ExternalDataProvider should log error message when user update method throws 1`] = `
[
  "The following error occured during Endpoint.update() for POST http://test.com/article-cooler",
]
`;

exports[`ExternalDataProvider should return denormalized value when schema is present (unions) 1`] = `
[
  [
    "Schema attribute "another" is not expected.
Expected one of: "first", "second"

Value: {
  "id": "6",
  "body": "hi",
  "type": "another"
}",
  ],
  [
    "Schema attribute undefined is not expected.
Expected one of: "first", "second"

Value: {
  "id": "7",
  "body": "hi"
}",
  ],
  [
    "TypeError: Unable to determine schema for UnionSchema
Value: {
  "id": "6",
  "body": "hi",
  "type": "another"
}.",
  ],
  [
    "TypeError: Unable to determine schema for UnionSchema
Value: {
  "id": "7",
  "body": "hi"
}.",
  ],
]
`;
